src/thunderbots/software/geom/util.cpp

```diff
@@ -0,0 +1,965 @@
+{
+    return (first.origin - second.origin).len() < (first.radius + second.radius);
+}
+
+bool intersects(const Ray &first, const Seg &second)
+{
+    if (std::abs(first.toVector().cross(second.toVector())) > EPS)
+    {
+        Vector isect = lineIntersection(first.start, first.dir, second.start, second.end);
+        return contains(first, isect) && contains(second, isect);
+    }
+    return collinear(first.start, first.dir, second.start);
+}
+bool intersects(const Seg &first, const Ray &second)
+{
+    return intersects(second, first);
+}
+
+bool intersects(const Seg &first, const Circle &second)
+{
+    // if the segment is inside the circle AND at least one of the points is
+    // outside the circle
+    return contains(second, first) &&
+           (distsq(first.start, second.origin) > second.radius * second.radius ||
+            distsq(first.end, second.origin) > second.radius * second.radius);
+}
+bool intersects(const Circle &first, const Seg &second)
+{
+    return intersects(second, first);
+}
+
+bool intersects(const Seg &first, const Seg &second)
+{
+    if (sign((first.start - first.end).cross(second.start - second.end)) == 0)
+    {
+        // find distance of two endpoints on segments furthest away from each
+        // other
+        double mx_len = std::sqrt(std::max(
+            std::max(
+                (second.start - first.end).lensq(), (second.end - first.end).lensq()),
+            std::max(
+                (second.start - first.start).lensq(),
+                (second.end - first.start).lensq())));
+        // if the segments cross then this distance should be less than
+        // the sum of the distances of the line segments
+        return mx_len <
+               (first.start - first.end).len() + (second.start - second.end).len() + EPS;
+    }
+
+    return sign((first.end - first.start).cross(second.start - first.start)) *
+                   sign((first.end - first.start).cross(second.end - first.start)) <=
+               0 &&
+           sign((second.end - second.start).cross(first.start - second.start)) *
+                   sign((second.end - second.start).cross(first.end - second.start)) <=
+               0;
+}
+
+template <size_t N>
+Vector getVertex(const Poly<N> &poly, unsigned int i)
+{
+    return poly[i % N];
+}
+
+template <size_t N>
+void setVertex(Poly<N> &poly, unsigned int i, const Vector &v)
+{
+    poly[i % N] = v;
+}
+
+template <size_t N>
+Seg getSide(const Poly<N> &poly, unsigned int i)
+{
+    return Seg(getVertex(poly, i), getVertex(poly, i + 1));
+}
+
+std::vector<std::pair<Vector, Angle>> angleSweepCirclesAll(
+    const Vector &src, const Vector &p1, const Vector &p2,
+    const std::vector<Vector> &obstacles, const double &radius)
+{
+    std::vector<std::pair<Vector, Angle>> ret;
+
+    const Angle offangle = (p1 - src).orientation();
+    if (collinear(src, p1, p2))
+    {
+        // std::cerr << "geom: collinear " << src << " " << p1 << " " << p2 <<
+        // std::endl;
```

Comment:
We should have better error handling here. We should decide if when the lines are colinear, if we just take the circles that intersect the line, or if we return an "empty" result. We can use std::optional for the empty result.

My vote would be for returning the circles that intersect the line.

Reply:
Could you clarify what you mean by that? I think it would make sense to return a vector containing the direction of the line and a zero radian or EPS radian angle if the line does not intersect any obstacles.

Reply:
According to the existing documentation, the function actually returns pairs of direction vectors and angles from said vectors that that are clear of obstacles between src and the line segment between p1 and p2 (ie. directions and angles that aren't blocked between a ball and the goal). I think it would make sense to return the direction and a zero angle if the line is not intersected by an obstacle.

Reply:
Ah yea I was misunderstanding the purpose of the function. According to the comment if the lines are colinear an empty vector is returned. I think what you said makes sense (return the line and Angle::zero()), so if you make that change make sure to update the comment.

Reply:
updated documentation